File: src/ainu_mcp/glossary.py
"""Google Sheets read/write for the Itak-uoeroskip glossary.

The sheet has one `all_sheets` meta tab listing content tabs, plus one tab per
semantic category. Each content tab is a table whose first row is column headers
(commonly: `Aynu`, `日本語`, `English`, `中文`, `註 / Notes`).

Behavior:
- Reads are cached per process for `_READ_TTL` seconds — search/list calls
  across many categories no longer hammer the 60-reads/min Sheets quota.
- `search` uses a single batchGet to fetch every category in one API call.
- 429s are retried with exponential backoff.
- Writes invalidate the cache for the affected tab so the next read is fresh.
- Updates support optimistic locking via a row hash.
"""

from __future__ import annotations

import hashlib
import random
import time
from dataclasses import dataclass
from functools import cache
from typing import Any, Callable, TypeVar

from google.oauth2 import service_account
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError

from .config import get_config

SCOPES = ["https://www.googleapis.com/auth/spreadsheets"]
_READ_TTL = 60.0  # seconds — short enough to pick up external edits, long enough to absorb burst reads

T = TypeVar("T")


@dataclass(frozen=True, slots=True)
class SheetMeta:
    is_content: bool
    sheet_name: str
    description: str
    count: int
    sheet_gid: int


@cache
def _sheets_service() -> Any:
    cfg = get_config()
    if cfg.google_credentials and cfg.google_credentials.exists():
        creds = service_account.Credentials.from_service_account_file(
            str(cfg.google_credentials), scopes=SCOPES
        )
    elif cfg.google_credentials_info:
        creds = service_account.Credentials.from_service_account_info(
            cfg.google_credentials_info, scopes=SCOPES
        )
    else:
        raise RuntimeError(
            "No Google credentials configured. Set GOOGLE_APPLICATION_CREDENTIALS "
            "to a service-account JSON path, or set the PRIVATE_GOOGLE_API_* env vars "
            "(compatible with the ainu-glossary site)."
        )
    return build("sheets", "v4", credentials=creds, cache_discovery=False)


def _row_hash(row: list[str]) -> str:
    """Stable 12-hex-char hash of a row's cell contents — used for optimistic locking."""
    joined = "".join(row)
    return hashlib.sha256(joined.encode("utf-8")).hexdigest()[:12]


def _col_letter(n: int) -> str:
    """1-indexed column number → spreadsheet letter (1→A, 27→AA)."""
    s = ""
    while n > 0:
        n, r = divmod(n - 1, 26)
        s = chr(65 + r) + s
    return s


def _with_retry(fn: Callable[[], T], *, retries: int = 5) -> T:
    """Run `fn`, retrying on transient 429/5xx with exponential backoff + jitter."""
    delay = 1.0
    for attempt in range(retries):
        try:
            return fn()
        except HttpError as e:
            status = getattr(e.resp, "status", None)
            if status in (429, 500, 502, 503, 504) and attempt < retries - 1:
                time.sleep(delay + random.uniform(0, 0.5))
                delay *= 2
                continue
            raise


# ── TTL cache (process-local) ─────────────────────────────────────────────── #

_cache_categories: tuple[float, list[dict[str, Any]]] | None = None
_cache_tabs: dict[str, tuple[float, tuple[list[str], list[list[str]]]]] = {}


def _invalidate(tab: str | None = None) -> None:
    """Drop the categories cache and (optionally) one tab's cache."""
    global _cache_categories
    _cache_categories = None
    if tab is None:
        _cache_tabs.clear()
    else:
        _cache_tabs.pop(tab, None)


def list_categories(*, force: bool = False) -> list[dict[str, Any]]:
    """Read the `all_sheets` tab and return the list of category sheets.

    Cached for `_READ_TTL` seconds. Pass `force=True` to bypass.
    """
    global _cache_categories
    now = time.monotonic()
    if not force and _cache_categories and now - _cache_categories[0] < _READ_TTL:
        return _cache_categories[1]

    svc = _sheets_service()
    cfg = get_config()
    resp = _with_retry(
        lambda: svc.spreadsheets()
        .values()
        .get(spreadsheetId=cfg.glossary_sheet_id, range="all_sheets")
        .execute()
    )
    rows = resp.get("values", [])
    out: list[dict[str, Any]] = []
    for r in rows[1:]:
        r = (r + [""] * 5)[:5]
        is_content, sheet_name, description, count, gid = r
        if not sheet_name:
            continue
        out.append(
            {
                "isContent": str(is_content).upper() == "TRUE",
                "sheetName": sheet_name,
                "description": description,
                "count": int(count) if str(count).strip().isdigit() else 0,
                "id": int(gid) if str(gid).strip().lstrip("-").isdigit() else None,
            }
        )
    _cache_categories = (now, out)
    return out


def _read_tab(sheet_name: str, *, force: bool = False) -> tuple[list[str], list[list[str]]]:
    """Return (headers, rows) for a content tab. Cached for `_READ_TTL` seconds."""
    now = time.monotonic()
    cached = _cache_tabs.get(sheet_name)
    if not force and cached and now - cached[0] < _READ_TTL:
        return cached[1]

    svc = _sheets_service()
    cfg = get_config()
    resp = _with_retry(
        lambda: svc.spreadsheets()
        .values()
        .get(spreadsheetId=cfg.glossary_sheet_id, range=f"'{sheet_name}'")
        .execute()
    )
    values = resp.get("values", [])
    if not values:
        result = ([], [])
    else:
        headers, *rows = values
        width = len(headers)
        rows = [(r + [""] * width)[:width] for r in rows]
        result = (headers, rows)
    _cache_tabs[sheet_name] = (now, result)
    return result


def _batch_read_tabs(sheet_names: list[str]) -> dict[str, tuple[list[str], list[list[str]]]]:
    """Fetch many tabs in one batchGet call (and cache them).

    Skips tabs already fresh in cache.
    """
    now = time.monotonic()
    needed = [
        n
        for n in sheet_names
        if not (
            (c := _cache_tabs.get(n)) and now - c[0] < _READ_TTL
        )
    ]
    if needed:
        svc = _sheets_service()
        cfg = get_config()
        resp = _with_retry(
            lambda: svc.spreadsheets()
            .values()
            .batchGet(
                spreadsheetId=cfg.glossary_sheet_id,
                ranges=[f"'{n}'" for n in needed],
            )
            .execute()
        )
        value_ranges = resp.get("valueRanges", [])
        for name, vr in zip(needed, value_ranges):
            values = vr.get("values", [])
            if not values:
                result = ([], [])
            else:
                headers, *rows = values
                width = len(headers)
                rows = [(r + [""] * width)[:width] for r in rows]
                result = (headers, rows)
            _cache_tabs[name] = (now, result)
    return {n: _cache_tabs[n][1] for n in sheet_names}


# ── Public surface ────────────────────────────────────────────────────────── #


def get_entry(category: str, row_number: int) -> dict[str, Any]:
    headers, rows = _read_tab(category)
    idx = row_number - 2
    if idx < 0 or idx >= len(rows):
        raise IndexError(f"row {row_number} out of range for '{category}' (has {len(rows)} data rows)")
    row = rows[idx]
    return {
        "category": category,
        "row": row_number,
        "row_hash": _row_hash(row),
        "fields": dict(zip(headers, row)),
    }


def list_entries(category: str, limit: int = 50, offset: int = 0) -> dict[str, Any]:
    headers, rows = _read_tab(category)
    sliced = rows[offset : offset + limit]
    entries = [
        {
            "category": category,
            "row": offset + 2 + i,
            "row_hash": _row_hash(r),
            "fields": dict(zip(headers, r)),
        }
        for i, r in enumerate(sliced)
    ]
    return {
        "category": category,
        "headers": headers,
        "total": len(rows),
        "returned": len(entries),
        "offset": offset,
        "entries": entries,
    }


def search(
    query: str,
    *,
    fields: list[str] | None = None,
    category: str | None = None,
    limit: int = 50,
) -> list[dict[str, Any]]:
    q = query.lower().strip()
    if not q:
        return []
    cats = (
        [category]
        if category
        else [c["sheetName"] for c in list_categories() if c["isContent"]]
    )
    tab_data = _batch_read_tabs(cats)
    hits: list[dict[str, Any]] = []
    for cat in cats:
        headers, rows = tab_data.get(cat, ([], []))
        target_fields = fields or headers
        for i, row in enumerate(rows):
            for h, cell in zip(headers, row):
                if h not in target_fields:
                    continue
                if q in cell.lower():
                    hits.append(
                        {
                            "category": cat,
                            "row": i + 2,
                            "row_hash": _row_hash(row),
                            "fields": dict(zip(headers, row)),
                            "matched_in": h,
                        }
                    )
                    break
            if len(hits) >= limit:
                return hits
    return hits


def untranslated(
    category: str | None = None,
    *,
    langs: list[str] | None = None,
    require_aynu: bool = True,
    limit: int = 200,
) -> dict[str, Any]:
    """Find rows where any of `langs` is empty.

    Default `langs` is ['日本語', 'English', '中文']. By default, rows missing
    the Aynu cell are skipped (since translation needs a source).
    Returns rows grouped by category, with which columns are missing per row.
    """
    target_langs = langs or ["日本語", "English", "中文"]
    cats = (
        [category]
        if category
        else [c["sheetName"] for c in list_categories() if c["isContent"]]
    )
    tab_data = _batch_read_tabs(cats)

    by_category: dict[str, list[dict[str, Any]]] = {}
    total_missing = 0
    for cat in cats:
        headers, rows = tab_data.get(cat, ([], []))
        if not headers:
            continue
        relevant_langs = [l for l in target_langs if l in headers]
        if not relevant_langs:
            continue
        cat_hits: list[dict[str, Any]] = []
        for i, row in enumerate(rows):
            fields = dict(zip(headers, row))
            if require_aynu and not (fields.get("Aynu") or "").strip():
                continue
            missing = [l for l in relevant_langs if not (fields.get(l) or "").strip()]
            if not missing:
                continue
            cat_hits.append(
                {
                    "category": cat,
                    "row": i + 2,
                    "row_hash": _row_hash(row),
                    "missing": missing,
                    "fields": fields,
                }
            )
            total_missing += 1
            if total_missing >= limit:
                if cat_hits:
                    by_category[cat] = cat_hits
                return {"total": total_missing, "by_category": by_category}
        if cat_hits:
            by_category[cat] = cat_hits
    return {"total": total_missing, "by_category": by_category}


def add_entry(category: str, fields: dict[str, str]) -> dict[str, Any]:
    headers, rows = _read_tab(category)
    new_row = [fields.get(h, "") for h in headers]
    svc = _sheets_service()
    cfg = get_config()
    resp = _with_retry(
        lambda: svc.spreadsheets()
        .values()
        .append(
            spreadsheetId=cfg.glossary_sheet_id,
            range=f"'{category}'",
            valueInputOption="USER_ENTERED",
            insertDataOption="INSERT_ROWS",
            body={"values": [new_row]},
        )
        .execute()
    )
    _invalidate(category)
    updated_range = resp.get("updates", {}).get("updatedRange", "")
    new_row_num = len(rows) + 2
    return {
        "category": category,
        "row": new_row_num,
        "row_hash": _row_hash(new_row),
        "fields": dict(zip(headers, new_row)),
        "updated_range": updated_range,
    }


def update_entry(
    category: str,
    row_number: int,
    fields: dict[str, str],
    expected_row_hash: str | None = None,
) -> dict[str, Any]:
    # Always read fresh for updates so we hash against the truth.
    headers, rows = _read_tab(category, force=True)
    idx = row_number - 2
    if idx < 0 or idx >= len(rows):
        raise IndexError(f"row {row_number} out of range for '{category}'")
    current = rows[idx]
    if expected_row_hash and _row_hash(current) != expected_row_hash:
        raise RuntimeError(
            f"row {row_number} in '{category}' has changed since last read "
            f"(expected hash {expected_row_hash}, got {_row_hash(current)}). "
            "Re-read the entry before updating."
        )

    unknown = [k for k in fields if k not in headers]
    if unknown:
        raise ValueError(f"unknown columns for '{category}': {unknown}; headers are {headers}")

    new_row = [fields.get(h, current[i]) for i, h in enumerate(headers)]
    end_col = _col_letter(len(headers))
    rng = f"'{category}'!A{row_number}:{end_col}{row_number}"
    svc = _sheets_service()
    cfg = get_config()
    _with_retry(
        lambda: svc.spreadsheets()
        .values()
        .update(
            spreadsheetId=cfg.glossary_sheet_id,
            range=rng,
            valueInputOption="USER_ENTERED",
            body={"values": [new_row]},
        )
        .execute()
    )
    _invalidate(category)
    return {
        "category": category,
        "row": row_number,
        "row_hash": _row_hash(new_row),
        "fields": dict(zip(headers, new_row)),
    }

File: worker/src/tools/glossary.ts
/**
 * Glossary read/write over Google Sheets (port of ainu_mcp/glossary.py).
 *
 * Reads are cached per isolate for _READ_TTL ms (the original 60s window) so
 * multi-category search/list calls stay well under the Sheets 60-reads/min
 * quota; writes invalidate the affected tab. Updates use optimistic locking via
 * a 12-hex-char row hash.
 */
import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { Env, Props } from "../types.js";
import { batchGetValues, getValues, appendValues, updateValues, tabRange } from "../sheets.js";
import { requireOrgMember } from "../auth.js";
import { jsonResult, errorResult } from "./helpers.js";

const READ_TTL = 60_000; // ms

export interface Category {
  isContent: boolean;
  sheetName: string;
  description: string;
  count: number;
  id: number | null;
}

interface Tab {
  headers: string[];
  rows: string[][];
}

// ── Per-isolate TTL cache ──
let cacheCategories: { at: number; data: Category[] } | null = null;
const cacheTabs = new Map<string, { at: number; data: Tab }>();

export function invalidate(tab?: string): void {
  cacheCategories = null;
  if (tab === undefined) cacheTabs.clear();
  else cacheTabs.delete(tab);
}

export async function rowHash(row: string[]): Promise<string> {
  const buf = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(row.join("")));
  return [...new Uint8Array(buf)].map((b) => b.toString(16).padStart(2, "0")).join("").slice(0, 12);
}

function colLetter(n: number): string {
  let s = "";
  while (n > 0) {
    const r = (n - 1) % 26;
    s = String.fromCharCode(65 + r) + s;
    n = Math.floor((n - 1) / 26);
  }
  return s;
}

export async function listCategories(env: Env, force = false): Promise<Category[]> {
  const now = Date.now();
  if (!force && cacheCategories && now - cacheCategories.at < READ_TTL) return cacheCategories.data;
  const rows = await getValues(env, "all_sheets");
  const out: Category[] = [];
  for (const raw of rows.slice(1)) {
    const r = [...raw, "", "", "", "", ""].slice(0, 5);
    const [isContent, sheetName, description, count, gid] = r;
    if (!sheetName) continue;
    out.push({
      isContent: String(isContent).toUpperCase() === "TRUE",
      sheetName,
      description,
      count: /^\d+$/.test(String(count).trim()) ? parseInt(count, 10) : 0,
      id: /^-?\d+$/.test(String(gid).trim()) ? parseInt(gid, 10) : null,
    });
  }
  cacheCategories = { at: now, data: out };
  return out;
}

export async function contentSheetNames(env: Env): Promise<string[]> {
  return (await listCategories(env)).filter((c) => c.isContent).map((c) => c.sheetName);
}

function toTab(values: string[][]): Tab {
  if (values.length === 0) return { headers: [], rows: [] };
  const [headers, ...rows] = values;
  const width = headers.length;
  const padded = rows.map((r) => [...r, ...Array(width).fill("")].slice(0, width));
  return { headers, rows: padded };
}

export async function readTab(env: Env, sheetName: string, force = false): Promise<Tab> {
  const now = Date.now();
  const cached = cacheTabs.get(sheetName);
  if (!force && cached && now - cached.at < READ_TTL) return cached.data;
  const values = await getValues(env, tabRange(sheetName));
  const tab = toTab(values);
  cacheTabs.set(sheetName, { at: now, data: tab });
  return tab;
}

export async function batchReadTabs(env: Env, sheetNames: string[]): Promise<Map<string, Tab>> {
  const now = Date.now();
  const needed = sheetNames.filter((n) => {
    const c = cacheTabs.get(n);
    return !(c && now - c.at < READ_TTL);
  });
  if (needed.length) {
    const valueRanges = await batchGetValues(env, needed.map(tabRange));
    needed.forEach((name, i) => {
      cacheTabs.set(name, { at: now, data: toTab(valueRanges[i] ?? []) });
    });
  }
  const result = new Map<string, Tab>();
  for (const n of sheetNames) result.set(n, cacheTabs.get(n)?.data ?? { headers: [], rows: [] });
  return result;
}

function fieldsOf(headers: string[], row: string[]): Record<string, string> {
  const o: Record<string, string> = {};
  headers.forEach((h, i) => (o[h] = row[i] ?? ""));
  return o;
}

// ──────────────────────────── Read tools ──────────────────────────── //

export function registerGlossaryReadTools(server: McpServer, env: Env): void {
  server.tool(
    "glossary_list_categories",
    "List every category (sheet tab) in the Itak-uoeroskip glossary, including metadata (description, entry count, sheet gid). Use this first to discover what categories exist before searching or adding entries.",
    {},
    async () => jsonResult(await listCategories(env)),
  );

  server.tool(
    "glossary_list_entries",
    "Page through entries in a category. Returns headers, total row count, and a slice of entries (each with row number and row_hash for safe editing).",
    { category: z.string(), limit: z.number().int().default(50), offset: z.number().int().default(0) },
    async ({ category, limit, offset }) => {
      const { headers, rows } = await readTab(env, category);
      const sliced = rows.slice(offset, offset + limit);
      const entries = await Promise.all(
        sliced.map(async (r, i) => ({
          category,
          row: offset + 2 + i,
          row_hash: await rowHash(r),
          fields: fieldsOf(headers, r),
        })),
      );
      return jsonResult({ category, headers, total: rows.length, returned: entries.length, offset, entries });
    },
  );

  server.tool(
    "glossary_get_entry",
    "Fetch one glossary entry by category + 1-indexed sheet row (row 1 is the header, so the first data row is 2). Returns fields and a row_hash to pass to glossary_update_entry for optimistic-locking safety.",
    { category: z.string(), row: z.number().int() },
    async ({ category, row }) => {
      const { headers, rows } = await readTab(env, category);
      const idx = row - 2;
      if (idx < 0 || idx >= rows.length) {
        return errorResult(`row ${row} out of range for '${category}' (has ${rows.length} data rows)`);
      }
      const r = rows[idx];
      return jsonResult({ category, row, row_hash: await rowHash(r), fields: fieldsOf(headers, r) });
    },
  );

  server.tool(
    "glossary_search",
    "Substring-search the glossary. Optionally restrict to specific columns (fields) or one category. Returns matching entries with row numbers and row_hashes.",
    {
      query: z.string(),
      fields: z.array(z.string()).optional(),
      category: z.string().optional(),
      limit: z.number().int().default(50),
    },
    async ({ query, fields, category, limit }) => jsonResult(await glossarySearch(env, { query, fields, category, limit })),
  );

  server.tool(
    "glossary_untranslated",
    "Find rows where one or more target language columns (日本語/English/中文 by default) is empty. Rows missing the Aynu cell are skipped by default. Returns rows grouped by category with the missing columns per row — a translation worklist.",
    {
      category: z.string().optional(),
      langs: z.array(z.string()).optional(),
      require_aynu: z.boolean().default(true),
      limit: z.number().int().default(200),
    },
    async ({ category, langs, require_aynu, limit }) =>
      jsonResult(await glossaryUntranslated(env, { category, langs, requireAynu: require_aynu, limit })),
  );
}

export async function glossarySearch(
  env: Env,
  opts: { query: string; fields?: string[]; category?: string; limit: number },
): Promise<Array<{ category: string; row: number; row_hash: string; fields: Record<string, string>; matched_in: string }>> {
  const q = opts.query.trim().toLowerCase();
  if (!q) return [];
  const cats = opts.category ? [opts.category] : await contentSheetNames(env);
  const tabData = await batchReadTabs(env, cats);
  const hits: Array<{ category: string; row: number; rowArr: string[]; headers: string[]; matched_in: string }> = [];
  for (const cat of cats) {
    const { headers, rows } = tabData.get(cat) ?? { headers: [], rows: [] };
    const targetFields = opts.fields ?? headers;
    for (let i = 0; i < rows.length; i++) {
      const row = rows[i];
      for (let h = 0; h < headers.length; h++) {
        if (!targetFields.includes(headers[h])) continue;
        if ((row[h] ?? "").toLowerCase().includes(q)) {
          hits.push({ category: cat, row: i + 2, rowArr: row, headers, matched_in: headers[h] });
          break;
        }
      }
      if (hits.length >= opts.limit) {
        return Promise.all(
          hits.map(async (hh) => ({
            category: hh.category,
            row: hh.row,
            row_hash: await rowHash(hh.rowArr),
            fields: fieldsOf(hh.headers, hh.rowArr),
            matched_in: hh.matched_in,
          })),
        );
      }
    }
  }
  return Promise.all(
    hits.map(async (hh) => ({
      category: hh.category,
      row: hh.row,
      row_hash: await rowHash(hh.rowArr),
      fields: fieldsOf(hh.headers, hh.rowArr),
      matched_in: hh.matched_in,
    })),
  );
}

async function glossaryUntranslated(
  env: Env,
  opts: { category?: string; langs?: string[]; requireAynu: boolean; limit: number },
) {
  const targetLangs = opts.langs ?? ["日本語", "English", "中文"];
  const cats = opts.category ? [opts.category] : await contentSheetNames(env);
  const tabData = await batchReadTabs(env, cats);
  const byCategory: Record<string, Array<{ category: string; row: number; row_hash: string; missing: string[]; fields: Record<string, string> }>> = {};
  let totalMissing = 0;
  for (const cat of cats) {
    const { headers, rows } = tabData.get(cat) ?? { headers: [], rows: [] };
    if (!headers.length) continue;
    const relevant = targetLangs.filter((l) => headers.includes(l));
    if (!relevant.length) continue;
    const catHits: Array<{ category: string; row: number; row_hash: string; missing: string[]; fields: Record<string, string> }> = [];
    for (let i = 0; i < rows.length; i++) {
      const fields = fieldsOf(headers, rows[i]);
      if (opts.requireAynu && !(fields["Aynu"] ?? "").trim()) continue;
      const missing = relevant.filter((l) => !(fields[l] ?? "").trim());
      if (!missing.length) continue;
      catHits.push({ category: cat, row: i + 2, row_hash: await rowHash(rows[i]), missing, fields });
      totalMissing++;
      if (totalMissing >= opts.limit) {
        if (catHits.length) byCategory[cat] = catHits;
        return { total: totalMissing, by_category: byCategory };
      }
    }
    if (catHits.length) byCategory[cat] = catHits;
  }
  return { total: totalMissing, by_category: byCategory };
}

// ──────────────────────────── Write tools ──────────────────────────── //

export function registerGlossaryWriteTools(server: McpServer, env: Env, props: Props): void {
  server.tool(
    "glossary_add_entry",
    "Append a new entry to a category tab. fields keys must match the tab's column headers (e.g. Aynu, 日本語, English, 中文, 註 / Notes); unknown keys are silently ignored. Returns the new row number and row_hash.",
    { category: z.string(), fields: z.record(z.string()) },
    async ({ category, fields }) => {
      requireOrgMember(props, "glossary_add_entry");
      const { headers, rows } = await readTab(env, category);
      const newRow = headers.map((h) => fields[h] ?? "");
      const { updatedRange } = await appendValues(env, tabRange(category), newRow);
      invalidate(category);
      return jsonResult({
        category,
        row: rows.length + 2,
        row_hash: await rowHash(newRow),
        fields: fieldsOf(headers, newRow),
        updated_range: updatedRange,
      });
    },
  );

  server.tool(
    "glossary_update_entry",
    "Update specific cells in an existing row. If expected_row_hash is supplied (recommended — from glossary_get_entry/search) and the row changed in the sheet since you read it, the update is refused so you can re-read and re-decide.",
    {
      category: z.string(),
      row: z.number().int(),
      fields: z.record(z.string()),
      expected_row_hash: z.string().optional(),
    },
    async ({ category, row, fields, expected_row_hash }) => {
      requireOrgMember(props, "glossary_update_entry");
      const { headers, rows } = await readTab(env, category, true); // read fresh for updates
      const idx = row - 2;
      if (idx < 0 || idx >= rows.length) return errorResult(`row ${row} out of range for '${category}'`);
      const current = rows[idx];
      if (expected_row_hash) {
        const actual = await rowHash(current);
        if (actual !== expected_row_hash) {
          return errorResult(
            `row ${row} in '${category}' has changed since last read (expected hash ${expected_row_hash}, got ${actual}). Re-read the entry before updating.`,
          );
        }
      }
      const unknown = Object.keys(fields).filter((k) => !headers.includes(k));
      if (unknown.length) {
        return errorResult(`unknown columns for '${category}': ${JSON.stringify(unknown)}; headers are ${JSON.stringify(headers)}`);
      }
      const newRow = headers.map((h, i) => (h in fields ? fields[h] : current[i] ?? ""));
      const endCol = colLetter(headers.length);
      const range = `${tabRange(category)}!A${row}:${endCol}${row}`;
      await updateValues(env, range, newRow);
      invalidate(category);
      return jsonResult({ category, row, row_hash: await rowHash(newRow), fields: fieldsOf(headers, newRow) });
    },
  );
}
